Sources/JSON.swift

```diff
@@ -14,10 +14,29 @@ enum JSONValueError: LocalizedError {
     }
 }
 
+public enum JSONNumber: Hashable {
+    case int(Int64)
+    case fraction(Double)
+    
+    public var intValue: Int? {
+        switch self {
+        case .int(let i): return Int(exactly: i)
+        case .fraction(let f): return Int(exactly: f)
```

Comment:
I wouldn't make this optional. I would make `intValue` truncate if it's actually a fraction.

Reply:
That's how NSNumber works at least, and I'd expect the same behavior here.

Reply:
that seems misinformative in this space though. JSON is a data transfer layer so i think it's important to maintain accuracy. i see `intValue` as more in use if the individual absolutely expects there to be an integer, in which case a `nil` value is an error and they can handle accordingly.

Reply:
To me we're talking about _decoding_ here though. Be incredibly rigorous with what you encode, but be incredibly forgiving with what you decode. If you want to use something as an int, you shouldn't have to rely on the server sending you `1` instead of `1.0` IMO. When you're encoding a JSONValue to be transferred, it totally makes sense to me to differentiate between the two and make sure you're capturing the intention correctly. Here I think it makes more sense to be flexible with what the server has given you.

Reply:
ah wait, what about overflow?

Reply:
What about it?

Reply:
how about i just convert to NSNumber and let it do it's thing? looks like that does not overflow, it just does bounds to max and min.
```swift
$ swift
Welcome to Apple Swift version 5.1.3 (swiftlang-1100.0.282.1 clang-1100.0.33.15).
Type :help for assistance.
  1> import Foundation
  2> let n  =  NSNumber(value: Int64.max) 
n: NSNumber = Int64(9223372036854775807)
  3> n.uint32Value
$R0: UInt32 = 4294967295
  4> UInt32.max
$R1: UInt32 = 4294967295
  5> let n2 = NSNumber(value: Int64.min)
n2: NSNumber = Int64(-9223372036854775808)
  6> n2.uint32Value
$R2: UInt32 = 0
  7>  
```

Reply:
/shrug whatever works for you as long as `intValue` isn't optional 😇

Reply:
>What about it?

it would crash the app
```swift
$ swift
Welcome to Apple Swift version 5.1.3 (swiftlang-1100.0.282.1 clang-1100.0.33.15).
Type :help for assistance.
  1> Int32(Int64.max)
Fatal error: Not enough bits to represent the passed value
2020-03-04 12:06:52.375376-0800 repl_swift[31220:19777194] Fatal error: Not enough bits to represent the passed value
Execution interrupted. Enter code to recover and continue.
Enter LLDB commands to investigate (type :help for assistance.)
Process 31220 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x00007fff6408d483 libswiftCore.dylib`function signature specialization <Arg[0] = Exploded, Arg[1] = Exploded, Arg[2] = Dead, Arg[3] = Dead> of Swift._fatalErrorMessage(_: Swift.StaticString, _: Swift.StaticString, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never + 307
libswiftCore.dylib`function signature specialization <Arg[0] = Exploded, Arg[1] = Exploded, Arg[2] = Dead, Arg[3] = Dead> of Swift._fatalErrorMessage(_: Swift.StaticString, _: Swift.StaticString, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never:
->  0x7fff6408d483 <+307>: ud2    
    0x7fff6408d485 <+309>: movq   %r12, %rdi
    0x7fff6408d488 <+312>: addq   $0x10, %rdi
    0x7fff6408d48c <+316>: cmpl   $0x80, %r14d
Target 0: (repl_swift) stopped.
  2>  
```